store/localstore/local_region.go

```diff
@@ -0,0 +1,318 @@
+	newStartKey []byte
+	newEndKey   []byte
+}
+
+func (rs *localRegion) Handle(req *regionRequest) (*regionResponse, error) {
+	resp := &regionResponse{
+		req: req,
+	}
+	if req.Tp == kv.ReqTypeSelect || req.Tp == kv.ReqTypeIndex {
+		sel := new(tipb.SelectRequest)
+		err := proto.Unmarshal(req.data, sel)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		txn := newTxn(rs.store, kv.Version{Ver: uint64(*sel.StartTs)})
+		var rows []*tipb.Row
+		if req.Tp == kv.ReqTypeSelect {
+			rows, err = rs.getRowsFromSelectReq(txn, sel)
+		} else {
+			rows, err = rs.getRowsFromIndexReq(txn, sel)
+		}
+		selResp := new(tipb.SelectResponse)
+		selResp.Error = toPBError(err)
+		selResp.Rows = rows
+		resp.err = err
+		data, err := proto.Marshal(selResp)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		resp.data = data
+	}
+	if bytes.Compare(rs.startKey, req.startKey) < 0 || bytes.Compare(rs.endKey, req.endKey) > 0 {
+		resp.newStartKey = rs.startKey
+		resp.newEndKey = rs.endKey
+	}
+	return resp, nil
+}
+
+func (rs *localRegion) getRowsFromSelectReq(txn kv.Transaction, sel *tipb.SelectRequest) ([]*tipb.Row, error) {
+	tid := sel.TableInfo.GetTableId()
+	kvRanges := rs.extractRecordKVRanges(tid, sel.Ranges)
+	var handles []int64
+	for _, ran := range kvRanges {
+		ranHandles, err := seekRangeHandles(tid, txn, ran)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		handles = append(handles, ranHandles...)
+	}
+	var rows []*tipb.Row
+	for _, handle := range handles {
+		row, err := rs.getRowFromHandle(txn, tid, handle, sel.TableInfo.Columns)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		rows = append(rows, row)
+	}
+	return rows, nil
+}
+
+func (rs *localRegion) extractRecordKVRanges(tid int64, krans []*tipb.KeyRange) []kv.KeyRange {
+	var kvRanges []kv.KeyRange
+	for _, kran := range krans {
+		upperKey := tablecodec.EncodeRowKey(tid, kran.GetHigh())
+		if bytes.Compare(upperKey, rs.startKey) <= 0 {
+			continue
+		}
+		lowerKey := tablecodec.EncodeRowKey(tid, kran.GetLow())
+		if bytes.Compare(lowerKey, rs.endKey) >= 0 {
+			continue
+		}
+		var kvr kv.KeyRange
+		if bytes.Compare(lowerKey, rs.startKey) <= 0 {
+			kvr.StartKey = rs.startKey
+		} else {
+			kvr.StartKey = lowerKey
+		}
+		if bytes.Compare(upperKey, rs.endKey) <= 0 {
+			kvr.EndKey = upperKey
+		} else {
+			kvr.EndKey = rs.endKey
+		}
+		kvRanges = append(kvRanges, kvr)
+	}
+	return kvRanges
+}
+
+func (rs *localRegion) getRowFromHandle(txn kv.Transaction, tid, handle int64, columns []*tipb.ColumnInfo) (*tipb.Row, error) {
+	row := new(tipb.Row)
+	var d types.Datum
+	d.SetInt64(handle)
+	var err error
+	row.Handle, err = codec.EncodeValue(nil, d)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, col := range columns {
+		if *col.PkHandle {
+			row.Data = append(row.Data, row.Handle...)
+		} else {
+			key := tablecodec.EncodeColumnKey(tid, handle, col.GetColumnId())
+			data, err := txn.Get(key)
```

Comment:
This will not called on real region server. Index data and row data maybe not on the same region server.